pallene/parser.lua

```diff
@@ -589,6 +562,39 @@ function Parser:FuncArgs()
     end
 end
 
+function Parser:FuncParams()
+    local oparen = self:e("(")
+    local params = self:DeclList()
+    local _ = self:e(")", oparen)
+    return params
+end
+
+function Parser:FuncExp(func_token, name_exp)
+    local params = self:FuncParams()
+    local rt_types = {}
+    if self:peek(":") then
+        self:region_begin()
+        self:e()
+        rt_types = self:RetTypes()
+        self:region_end()
+    end
+
+    local block    = self:Block()
+    local _        = self:e("end", func_token)
+
+    local arg_types = {}
+    for i, decl in ipairs(params) do
+        if not decl.type then
+            self:syntax_error(decl.loc,
+                    "Parameter '%s' is missing a type annotation", decl.name)
```

Comment:
We should consider removing this restriction from the parser. It might be useful to allow lambdas without a type annotation, if they are used in a context where the type is known:

    local f : integer->integer = function(x) return x+ 1 end
   
    local function map(xs: {any}, f: any->any): {any}
        -- ...
    end
    map(my_list, function(x) return x end)

That said, this change doesn't need to be right now.

Reply:
just to make sure, this is only for lambdas and not for function statements right?

Reply:
Yes. Perhaps that error checking shoulg be just on the fn statements.

Reply:
Instead of having an is_stat argument, we could put this error checking in the `Func` (which we should probably also rename to `FuncStat`, by the way). After the `FuncExp` returns, iterate over the returned expression to see if if all the decls have the required type annotation.